WIP API client implementation

kothar opened the pull request:
This isn't complete in any way, I've just started implementing some of the API based on your skeleton.

I thought I'd start a pull request now to avoid duplicated effort if/when you are also working on this.

Regards,
Mike.

kothar committed:
Implemented the b2_authorize_account call

kothar committed:
Preliminary work on a basic b2 command line client

kothar committed:
File upload implemented, but getting a 500 status error with no explanation

kothar committed:
List and Put commands work

Put requires the whole file to be buffered for some reason - os.File doesn't seem to work,
while bytes.Reader does.

kothar committed:
Implemented file download support

kothar committed:
Refactored code into bucket and file-grouped methods

tblyler commented:
I'm not sure if we want to combine efforts, but I have a nearly complete API for this. https://github.com/tblyler/go-blaze
I primarily just need to get testing finished and possibly rearrange the structs a bit better.

kothar commented:
@tblyler Great, I thought some other people were probably already working on this, but this repo was the only one I found with a brief search. I think I'm nearly done too, so we could compare implementations and see if it's worth merging anything.

kothar committed:
Remaining API requests implemented

kothar committed:
Implemented delete command in CLI

kothar committed:
Implemented createbucket and deletebucket commands

Improved documentation for b2 client

kothar committed:
Add metadata when uploading files

kothar committed:
Buffer file before uploading

Uploading a file to B2 from os.File directly cause internal errors, so buffer while
computing the SHA1 hash.

kothar committed:
Fixes for progress feedback when seeking and B2 internal errors

TeeReader is not seekable, so can't be used as a viable input for upload if progress
reporting is desired.

Internal errors are caused if the Content-Length header is not provided. This was only being set
if the request body was a Buffer or ByteReader.
